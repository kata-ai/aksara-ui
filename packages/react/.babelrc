{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "browsers": ["last 2 chrome versions", "last 2 firefox versions"]
        }
      }
    ],
    "@babel/preset-react",
    "@babel/preset-typescript"
  ],
  "plugins": [
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-optional-chaining",
    "@babel/plugin-proposal-nullish-coalescing-operator",
    "babel-plugin-styled-components"
  ],
  "env": {
    "test": {
      "plugins": ["@babel/plugin-transform-modules-commonjs"]
    }
  }
}
