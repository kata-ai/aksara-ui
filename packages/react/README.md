# @aksara-ui/react

> React implementation of Aksara Design System, from Kata.ai.

This package is part of the Aksara monorepo. Please read the README in [the Aksara repository](https://github.com/kata-ai/aksara-ui) for detailed documentation about this package.
