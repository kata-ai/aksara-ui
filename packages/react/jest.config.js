module.exports = {
  collectCoverageFrom: [
    'src/**/*.{ts,tsx}',
    '!src/**/*.stories.{ts,tsx}',
    '!src/**/*.story.{ts,tsx}',
    '!src/**/index.{ts,tsx}',
    '!src/stories/**/*.{ts,tsx}',
    '!src/utils/storybook/**/*.{ts,tsx}',
    '!src/index.{ts,tsx}',
  ],
  coverageReporters: ['json', 'lcov'],
  moduleDirectories: ['node_modules', '<rootDir>/src'],
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
  modulePathIgnorePatterns: ['lib', 'lib-es6'],
  setupFilesAfterEnv: ['<rootDir>/test/setupTests.ts'],
  testPathIgnorePatterns: ['<rootDir>/dist/'],
  transform: {
    '^.+\\.(js|jsx|ts|tsx|mjs)$': 'babel-jest',
    '^.+\\.css$': '<rootDir>/config/jest/cssTransform.js',
    '^(?!.*\\.(js|jsx|ts|tsx|mjs|css|json)$)': '<rootDir>/config/jest/fileTransform.js',
  },
};
