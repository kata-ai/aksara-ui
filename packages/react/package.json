{
  "name": "@aksara-ui/react",
  "version": "1.0.1-canary.4",
  "description": "React implementation of Aksara Design System, from Kata.ai.",
  "main": "dist/index.js",
  "module": "dist/core.esm.js",
  "typings": "dist/index.d.ts",
  "aksara:src": "./src/index.ts",
  "repository": "https://github.com/kata-ai/wicara.git",
  "author": "Kata.ai <dev@kata.ai>",
  "license": "Apache-2.0",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsdx build --transpileOnly --tsconfig tsconfig.build.json",
    "build-storybook": "build-storybook --quiet",
    "clean": "rimraf dist",
    "format": "prettier --write \"src/**/*.{js,jsx,ts,tsx}\"",
    "lint": "eslint \"src/**/*.{js,jsx,ts,tsx}\"",
    "prepublish": "yarn build",
    "report-coverage": "codecov -f ./coverage/*.json",
    "start": "tsdx watch --tsconfig tsconfig.build.json --verbose",
    "storybook": "start-storybook -p 9001",
    "storybook:build": "build-storybook --quiet",
    "test": "jest",
    "test:coverage": "jest --coverage",
    "type-check": "tsc --emitDeclarationOnly"
  },
  "dependencies": {
    "@aksara-ui/core": "^1.0.1-canary.1",
    "@popperjs/core": "^2.10.2",
    "@reach/accordion": "^0.16.1",
    "@reach/visually-hidden": "^0.16.0",
    "@styled-system/css": "^5.1.5",
    "@styled-system/should-forward-prop": "^5.1.5",
    "@styled-system/theme-get": "^5.1.2",
    "@types/styled-system": "^5.1.13",
    "@types/styled-system__css": "^5.0.16",
    "@types/styled-system__theme-get": "^5.0.1",
    "clsx": "^1.1.1",
    "csstype": "^3.0.9",
    "deepmerge": "^4.2.2",
    "focus-trap": "^6.6.0",
    "polished": "^4.1.3",
    "react-fast-compare": "^3.2.0",
    "react-popper": "^2.2.5",
    "react-popper-tooltip": "^4.3.0",
    "react-transition-group": "^4.4.2",
    "styled-system": "^5.1.4",
    "tslib": "^2.3.1"
  },
  "peerDependencies": {
    "@aksara-ui/icons": "^1.0.0-canary.0",
    "react": "^16.10.0 || ^17.0.0",
    "react-dom": "^16.10.0 || ^17.0.0",
    "styled-components": "^5.1.0"
  },
  "devDependencies": {
    "@aksara-ui/icons": "^1.0.0",
    "@babel/core": "^7.15.0",
    "@storybook/addon-actions": "^6.3.7",
    "@storybook/addon-essentials": "^6.3.7",
    "@storybook/addon-knobs": "^6.3.0",
    "@storybook/addon-links": "^6.3.7",
    "@storybook/react": "^6.3.7",
    "@storybook/theming": "^6.3.7",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^12.0.0",
    "@types/jest": "^26.0.24",
    "@types/react-transition-group": "^4.4.2",
    "@types/styled-components": "^5.1.12",
    "babel-loader": "^8.2.2",
    "babel-plugin-styled-components": "^1.13.2",
    "babel-preset-react-app": "^10.0.0",
    "codecov": "^3.8.3",
    "copee": "^1.0.6",
    "jest": "^26.6.3",
    "jest-styled-components": "^7.0.5",
    "react": "^17.0.2",
    "react-docgen-typescript-loader": "^3.6.0",
    "react-dom": "^17.0.2",
    "styled-components": "^5.3.0"
  }
}
