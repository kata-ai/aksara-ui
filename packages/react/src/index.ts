export * from './aksara-provider';
export * from './components';
export * from './layout';
export * from './theme';
export * from './typography';
export * from '@aksara-ui/helpers';
