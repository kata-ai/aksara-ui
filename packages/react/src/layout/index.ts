export * from './box';
export * from './inline';
export * from './stack';
export * from './wrap';
