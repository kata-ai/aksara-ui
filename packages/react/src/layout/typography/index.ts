export * from './components/Anchor';
export * from './components/Heading';
export * from './components/Paragraph';
export * from './components/Text';
export * from './components/UnstyledAnchor';

export * from './utils';
