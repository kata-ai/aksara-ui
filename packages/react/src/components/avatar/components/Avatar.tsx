import * as React from 'react';
import { avatarSizeOptions } from '../../../theme/component-styles/avatar';
import { VisuallyHidden } from '../../../helpers';
import { Box, BoxProps } from '../../../layout';
import { useComponentStyles } from '../../../system';
import { Text } from '../../../typography';
import getInitials from '../utils/getInitials';
import Presence, { PresenceProps } from './Presence';

type AvatarSize = 'sm' | 'md' | 'lg' | 'xl';
const bgColors = ['indigo06', 'red07', 'yellow07', 'blue07', 'green07'];
export interface AvatarProps extends Omit<BoxProps, 'size'> {
  /** Additional CSS classes to give to the component. */
  className?: string;
  /** Additional CSS styles to give to the component. */
  style?: React.CSSProperties;
  /** Name for initials */
  name?: string;
  /** The avatar's image source. */
  src?: any;
  /** Alt text for the avatar. */
  alt?: string;
  /** The background color of the avatar. */
  bg?: string;
  /** The icon that renders with the avatar. */
  icon?: React.ComponentType<any>;
  /** Size of the avatar. */
  size?: AvatarSize;
  /** Presence option */
  presence?: PresenceProps;
}

function iconSizes(size: AvatarSize = 'lg') {
  return avatarSizeOptions[size] - 8;
}

const initialSizes: Record<AvatarSize, string> = {
  sm: '8px',
  md: '12px',
  lg: '14px',
  xl: '20px',
};

/** Resizable avatar component. */
const Avatar = React.forwardRef<HTMLSpanElement, AvatarProps>(
  (
    { className, style, src, alt = undefined, name, size = 'lg', color = 'white', bg, icon, presence, ...rest },
    ref
  ) => {
    const styles = useComponentStyles('avatar', { size: typeof size === 'string' ? size : undefined });

    const renderInitials = () => {
      return (
        <Text fontSize={initialSizes[size]} lineHeight={1} role="presentation">
          {name ? getInitials(name) : '??'}
        </Text>
      );
    };

    const renderImage = () => {
      if (src) {
        return (
          <Box
            as="img"
            sx={{ position: 'absolute', width: '100%', height: '100%', objectFit: 'cover' }}
            src={src}
            alt={name || alt}
          />
        );
      }

      if (icon) {
        return React.createElement(icon, { fill: 'currentColor', size: iconSizes(size), 'aria-hidden': true });
      }

      return renderInitials();
    };

    const renderPresence = (option: PresenceProps) => {
      return <Presence {...option} size={size} />;
    };

    const bgColorsRandomizer = () => {
      if (!name) {
        return 'indigo06';
      }
      const names = name.split(' ');
      let totalValue = 0;
      names.forEach(item => {
        const utfValue = new TextEncoder().encode(`${item[0]}${item[item.length - 1]}`);
        totalValue += utfValue[0] + utfValue[1];
      });
      const indexColor = totalValue % bgColors.length;
      return bgColors[indexColor];
    };

    const renderAvatar = () => {
      return (
        <Box
          data-testid="avatar-container"
          as="span"
          ref={ref}
          className={className}
          style={style}
          color={color}
          bg={bg || bgColorsRandomizer()}
          sx={{ ...styles }}
          width={typeof size === 'number' ? size : undefined}
          height={typeof size === 'number' ? size : undefined}
          borderRadius={9999}
          {...rest}
        >
          {name && <VisuallyHidden>{name}</VisuallyHidden>}
          {renderImage()}
        </Box>
      );
    };
    if (presence) {
      return (
        <Box position="relative" data-testid="avatar-with-presence">
          {renderAvatar()}
          {renderPresence(presence)}
        </Box>
      );
    }
    return renderAvatar();
  }
);

Avatar.displayName = 'Avatar';

export default Avatar;
