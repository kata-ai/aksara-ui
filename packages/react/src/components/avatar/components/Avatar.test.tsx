import * as React from 'react';
import { render } from '@testing-library/react';

import Avatar from './Avatar';
import { Box } from '../../../layout';

describe('components/Avatar', () => {
  describe('<Avatar />', () => {
    test('renders image correctly', () => {
      const { getByRole } = render(<Avatar name="Adry Muhammad" src="https://picsum.photos/id/2/400/400" />);

      const image = getByRole('img', {
        name: /adry muhammad/i,
      });
      expect(image).toBeVisible();
    });

    test('renders initials correctly', () => {
      const { getByText, getByRole } = render(<Avatar name="Adry Muhammad" />);

      expect(getByText(/adry muhammad/i)).toBeInTheDocument();
      expect(getByRole('presentation')).toBeVisible();
    });

    test('render presence and notification badge correctly', () => {
      const { getByTestId } = render(
        <Avatar
          name="Adry Muhammad"
          presence={<Box width="100%" height="100%" backgroundColor={'red07'} />}
          notificationBadge="99+"
        />
      );
      const presence = getByTestId('sign-badge');
      expect(presence).toBeInTheDocument();
      const notificationBadge = getByTestId('notification-badge');
      expect(notificationBadge).toBeInTheDocument();
    });

    test('not render presence', () => {
      const { queryByTestId } = render(<Avatar name="Adry Muhammad" />);
      const presence = queryByTestId('sign-badge');
      expect(presence).not.toBeInTheDocument();
      const notificationBadge = queryByTestId('notification-badge');
      expect(notificationBadge).not.toBeInTheDocument();
    });
  });
  describe('Bg Color Generator', () => {
    test('generate bgColor Correctly', () => {
      const { getAllByTestId } = render(
        <>
          <Avatar name="Adry Baba" />
          <Avatar name="Rara Permai Sari" />
          <Avatar name="Ardy Buba" />
          <Avatar name="Ardy Brai" />
        </>
      );

      const listContainer = getAllByTestId('avatar-container');
      expect(!!(listContainer[0].className === listContainer[2].className)).toBe(true);
      expect(!!(listContainer[0].className === listContainer[1].className)).toBe(false);
      expect(!!(listContainer[0].className === listContainer[3].className)).toBe(false);
    });
  });
});
