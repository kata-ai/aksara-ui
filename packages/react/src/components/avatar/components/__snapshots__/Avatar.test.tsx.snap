// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Avatar <Avatar /> renders in correct sizes 1`] = `
.c0 {
  color: white;
  background-color: indigo05;
  width: 24px;
  height: 24px;
  border-radius: 24px;
}

<span
  class="c0"
>
  <span
    class="Text-sc-1e9atua-0"
  >
    ??
  </span>
</span>
`;

exports[`components/Avatar <Avatar /> renders on large size by default 1`] = `
.c0 {
  color: white;
  background-color: indigo05;
  width: 40px;
  height: 40px;
  border-radius: 40px;
}

<span
  class="c0"
>
  <span
    class="Text-sc-1e9atua-0"
  >
    ??
  </span>
</span>
`;
